File: RESTFul/SpeechCustom/PHP/app1/lib/Speech/NBest.php
<?php
namespace Att\Api\Speech;

/*
 * Copyright 2014 AT&T
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Immutable class that represents the N-Best hypothesis for speech recognization.
 *
 * @category API
 * @package  Speech
 * @author   pk9069
 * @license  http://www.apache.org/licenses/LICENSE-2.0
 * @version  Release: @package_version@ 
 * @link     https://developer.att.com/docs/apis/rest/3/Speech
 */
final class NBest
{
    /**
     * hypothesis.
     *
     * @var string
     */
    private $_hypothesis;

    /**
     * Language id used to decode speech.
     *
     * @var string
     */
    private $_languageId;

    /**
     * Confidence level value, where value: 0 <= value <= 1.0
     *
     * @var string
     */
    private $_confidence;

    /**
     * Assessment of quality.
     *
     * @var string
     */
    private $_grade;

    /**
     * Result text.
     *
     * @var string
     */
    private $_resultText;
    
    /**
     * Words recognized from speech.
     *
     * @var array 
     */
    private $_words;

    /**
     * Confidence score for each word.
     *
     * @var array
     */
    private $_wordScores;

    /**
     * Creates an N-Best hypothesis object.
     *
     * @param string $hypothesis hypothesis for speech recognization
     * @param string $languageId language Id
     * @param string $confidence confidence
     * @param string $grade      grade
     * @param string $resultText result text
     * @param array  $words      array of words
     * @param array  $wordScores array of word scores
     */
    public function __construct($hypothesis, $languageId, $confidence, $grade,
        $resultText, $words, $wordScores
    ) {
        $this->_hypothesis = $hypothesis;
        $this->_languageId = $languageId;
        $this->_confidence = $confidence;
        $this->_grade = $grade;
        $this->_resultText = $resultText;
        $this->_words = $words;
        $this->_wordScores = $wordScores;
    }

    /**
     * Gets the hypothesis.
     *
     * @return string hypothesis
     */
    public function getHypothesis()
    {
        return $this->_hypothesis;
    }

    /**
     * Gets the language ID.
     *
     * @return language Id
     */
    public function getLanguageId()
    {
        return $this->_languageId;
    }

    /**
     * Gets the confidence for word recognization.
     *
     * @return string confidence
     */
    public function getConfidence()
    {
        return $this->_confidence;
    }

    /**
     * Gets grade.
     *
     * @return string Grade
     */
    public function getGrade()
    {
        return $this->_grade;
    }

    /**
     * Gets result text.
     *
     * @return string result text
     */
    public function getResultText()
    {
        return $this->_resultText;
    }

    /**
     * Gets an array of words.
     *
     * @return array words
     */
    public function getWords()
    {
        return $this->_words;
    }

    /**
     * Gets an array of word scores.
     *
     * @return array word scores
     */
    public function getWordScores()
    {
        return $this->_wordScores;
    }

    public static function fromArray($arr)
    {
        $nBest = new NBest(
            $arr['Hypothesis'],
            $arr['LanguageId'],
            $arr['Confidence'],
            $arr['Grade'],
            $arr['ResultText'],
            $arr['Words'],
            $arr['WordScores']
        );

        return $nBest;
    }
}

/* vim: set expandtab tabstop=4 shiftwidth=4 softtabstop=4: */
?>

File: RESTFul/Speech/Java/app1/src/main/java/com/att/api/speech/model/NBest.java
package com.att.api.speech.model;

import java.util.ArrayList;
import java.util.List;

import org.json.JSONArray;
import org.json.JSONObject;

public class NBest {
    private String hypothesis;
    private String languageId;
    private double confidence;
    private String grade;
    private String resultText;
    private NLUHypothesis nluHypothesis;
    private List<String> words;
    private List<Double> wordScores;

    private NBest(Builder builder) {
        this.hypothesis = builder.hypothesis;
        this.languageId = builder.languageId;
        this.confidence = builder.confidence;
        this.grade = builder.grade;
        this.resultText = builder.resultText;
        this.nluHypothesis = builder.nluHypothesis;
        this.words = builder.words;
        this.wordScores = builder.wordScores;
    }

    /**
     * @return the hypothesis
     */
    public String getHypothesis() {
        return this.hypothesis;
    }

    /**
     * @return the languageId
     */
    public String getLanguageId() {
        return this.languageId;
    }

    /**
     * @return the confidence
     */
    public double getConfidence() {
        return this.confidence;
    }

    /**
     * @return the grade
     */
    public String getGrade() {
        return this.grade;
    }

    /**
     * @return the resultText
     */
    public String getResultText() {
        return this.resultText;
    }

    /**
     * @return the nluHypothesis
     */
    public NLUHypothesis getNluHypothesis() {
        return this.nluHypothesis;
    }

    /**
     * @return the words
     */
    public List<String> getWords() {
        return this.words;
    }

    /**
     * @return the wordScores
     */
    public List<Double> getWordScores() {
        return this.wordScores;
    }

    public static List<String> jsonStringArrayToList(JSONArray arr) {
        List<String> list = new ArrayList<String>();
        for (int i = 0; i < arr.length(); ++i) {
            list.add(arr.getString(i));
        }
        return list;
    }

    public static List<Double> jsonDoubleArrayToList(JSONArray arr) {
        List<Double> list = new ArrayList<Double>();
        for (int i = 0; i < arr.length(); ++i) {
            list.add(arr.getDouble(i));
        }
        return list;
    }

    public static NBest valueOf(JSONObject json) {
        // Add optional parts
        Builder builder = new Builder();

        String key = "LanguageId";
        if (json.has(key))
                builder.setLanguageId(json.getString(key));

        key = "ResultText";
        if (json.has(key))
                builder.setResultText(json.getString(key));

        key = "Words";
        if (json.has(key)){
            List<String> words = jsonStringArrayToList(json.getJSONArray(key));
            builder.setWords(words);
        }

        key = "WordScores";
        if (json.has(key)){
            List<Double> scores = jsonDoubleArrayToList(json.getJSONArray(key));
            builder.setWordScores(scores);
        }

        key = "NluHypothesis";
        if (json.has(key)){
            NLUHypothesis nlu = NLUHypothesis.valueOf(json.getJSONObject(key));
            builder.setNluHypothesis(nlu);
        }

        // Add required parts and build new object
        return builder
            .setHypothesis(json.getString("Hypothesis"))
            .setConfidence(json.getDouble("Confidence"))
            .setGrade(json.getString("Grade"))
            .build();
    }

    public static class Builder {
        private String hypothesis;
        private String languageId;
        private double confidence;
        private String grade;
        private String resultText;
        private NLUHypothesis nluHypothesis;
        private List<String> words;
        private List<Double> wordScores;

        /**
         * @param hypothesis the hypothesis to set
         */
        public Builder setHypothesis(String hypothesis) {
            this.hypothesis = hypothesis;
            return this;
        }

        /**
         * @param languageId the languageId to set
         */
        public Builder setLanguageId(String languageId) {
            this.languageId = languageId;
            return this;
        }

        /**
         * @param confidence the confidence to set
         */
        public Builder setConfidence(double confidence) {
            this.confidence = confidence;
            return this;
        }

        /**
         * @param grade the grade to set
         */
        public Builder setGrade(String grade) {
            this.grade = grade;
            return this;
        }

        /**
         * @param resultText the resultText to set
         */
        public Builder setResultText(String resultText) {
            this.resultText = resultText;
            return this;
        }

        /**
         * @param nluHypothesis the nluHypothesis to set
         */
        public Builder setNluHypothesis(NLUHypothesis nluHypothesis) {
            this.nluHypothesis = nluHypothesis;
            return this;
        }

        /**
         * @param words the words to set
         */
        public Builder setWords(List<String> words) {
            this.words = words;
            return this;
        }

        /**
         * @param wordScores the wordScores to set
         */
        public Builder setWordScores(List<Double> wordScores) {
            this.wordScores = wordScores;
            return this;
        }

        public NBest build(){
            return new NBest(this);
        }
    }
}
